### DISCUSSION TITLE
Multi-AZ and failover

### DISCUSSION THREAD

**Discussion by solracsf** [juicedata/juicefs#1566] [Ideas]
JuiceFS already supports  one endpoint, let's say s3.us-east-1.amazonaws.com, by setting

`--bucket bucket.s3.us-east-1.amazonaws.com`

Problem here is that data will be stored in one region only `us-east-1`, so no Multi-AZ. Some providers allow bucket replication, but juiceFS wouldn't know how to reach them as only one endpoint is set... 

My idea here was that if one could use 2 (or more) endpoints, data would be somehow replicated in these 2 buckets, in different regions, and JuiceFS could always try bucket 2 in case bucket 1 request fails for whatever reason for READ (not sure for WRITES) statements.

Imagine one can set this:

```
--bucket bucket1.s3.us-east-1.amazonaws.com
--failover bucket2.s3.us-east-2.amazonaws.com
```

As they both are accessed with the same access_key/secret_key, data would be somehow replicated over 2 regions.

**WRITE** request:
- Write to `--bucket`
- In the background, Write to `--failover`

Problem here is how to deal with fail requests; if either --bucket or --failover is not reachable, there would be 2 options: 
- WRITE only in one of them, as READ would try `--bucket` first, followed by `--failover`, _see below_.
- WRITE in the available bucket, and replicate back to the other one in the background once it's available again (not sure if this could be done, and how).
_- Another approach?_

**READ** requests are more "easy" to handle:
- Try `--bucket`
- If it fails (4xx or 5xx), try `--failover`

`--failover` option would be available for new filesystems only, not existing ones.

I don't know if this makes sense, but it could provide a good failover system and better availability and durability.

**Comment by davies** [comment]
@solracsf  us-east-1 is a region with multiple available zones builtin, so any object written into us-east-1 is already replicated into multiple AZ by default, any failure of single AZ will not result in unavailability or data loss.

For replication across multiple regions or cloud vendors, is on the roadmap.
